### ISSUE TITLE
Unhandled exceptions take down unexpected functionality

### ISSUE THREAD

**Issue by Akin909** [onivim/oni#1488]
@bryphe  this is related somewhat to #1383 but I've noticed a pattern where I lose a certain bit of functionality sporadically whilst using oni i.e on startup or several hours later but in general a lot of these errors seem to propagate and take down some bit of functionality some where sometimes in a very unexpected way. I believe #1487 is one example, another is #1419.

Not sure what a good solution is but my brute force very inelegant knee jerk reaction is to wrap every function in the `NeovimEditor` for starters in try catches 😆

**Comment by bryphe** [comment]
> @bryphe this is related somewhat to #1383 but I've noticed a pattern where I lose a certain bit of functionality sporadically whilst using oni i.e on startup or several hours later but in general a lot of these errors seem to propagate and take down some bit of functionality some where sometimes in a very unexpected way.

Yes, definitely frustrating! We still have a ways to go in terms of building up 'robustness'

> Not sure what a good solution is but my brute force very inelegant knee jerk reaction is to wrap every function in the NeovimEditor for starters in try catches 😆

Haha, unfortunately I don't think this is a tenable solution (unfortunately)

The main issue that happens is that, once an unhandled exception occurs, often the app can be in an undefined state. Imagine you get an exception right before setting some critical piece of state - catching it might keep the app from crashing at that point, but now the state is inconsistent, and you might get even _crazier_ issues due to that inconsistent state. The impact of that is you can spend your time debugging lots of different issues that have a common root cause!

When you compare to a runtime like C++ or C#, if there is an unhandled exception, the app just crashes / quits! In some cases, that might be more desirable.

Our best defense here is to put the errors _front and center_ - we should let the user know as soon as we hit these issues, because once there is an unhandled exception, all bets are off. We should give them the tools to quickly log (ideally, we'd have telemetry too so we could keep an eye on the most common unhandled exceptions, and prioritize them!)

In parallel, we should try and catch these errors as quickly as possible. It can be challenging to test all the cases due to the variety of configurations & vim plugins, but  as we find issues, it's helpful to build a `CiTest` or something to reproduce it (like the regression test added in PR #1309 for the window-opening-crash). These tests pay dividends because we know that error is covered and won't come back!

Today our automation was swallowing errors... and a lot of errors if they didn't have user impact were sort of being swept under the rug... but I think if we make them "front and center" and also monitor them in automation (#1491), we can make a push forward in terms of robustness.

**Comment by Akin909** [comment]
Seems #1494 is a robust solution to this issue I'd be happy to close it if you don't want to use it to track against though its fairly nebulous so wouldn't make for a very good tracker

**Comment by bryphe** [comment]
> Seems #1494 is a robust solution to this issue I'd be happy to close it if you don't want to use it to track against though its fairly nebulous so wouldn't make for a very good tracker

Sounds great, thanks @Akin909 ! We can open a new issue if we find there are still classes of exceptions being swept under the rug, or other pain points causing issues.
